### PR TITLE
Avoid top-level `with ...;` in `pkgs/development/haskell-modules`

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Using `with` at a top-level scope is [an anti-pattern](https://nix.dev/guides/best-practices#with-scopes). The tracking issue is #208242. This is a pure refactor: there is no functional change contained in this PR.

I used the refactor strategy that looks for the uses of names [coming from `lib`, `pkgs`, and `haskellLib`](https://gist.github.com/philiptaron/04326000d5ff20cb72c2805efd09701c). 

<del>Following advice, I also preferred to inherit names from `lib` instead of `builtins` where there was a choice. I also preferred to inherit from `lib` in one place, rather than to have a mixture of `lib.thing`, `pkgs.lib.thing`, and `inherit (lib) thing`. **I can revert these changes if you all feel them to be net-negative.**</del> I have reverted them.

<del>There are no rebuilds when I run `nixpkgs-review rev HEAD`, so I've targeted master with this PR.</del> @maralorn asked me to target `haskell-updates`, so I have done so.

On a personal note, I don't particularly like the long list of inherited names in (say) `configuration-common.nix` but I think it's better than a top-level with statement.

## Things done

- [x] Tested compilation of all packages that depend on this change using `nixpkgs-review rev HEAD`.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by maralorn** [comment]
While the long import lists are annoying to book keep getting rid of the with statement is worth it. So thank you for that.

However most of the changes in this PR are converting `lib.foo` to `inherit (lib) foo`. Since this has nothing to do with `with` they need a different and stronger justification. Let’s collect some argument.

Pro:

* shorter code
* We already do this for some functions in some files in `haskell-modules`
* consistent way of importing things

Cons:

* harder to tell whether an imported library function is from lib or haskellLib.
* Increases "annoying bookkeeping" implying potential merge conflicts and jumping in the file during development.
* Adds noise to this PR, could be done separately

Maybe others have other arguments? I am on the fence.

Thank you anyway for pushing this. While resistance is to be expected I appreciate that people are trying to work on nixpkgs code quality.

**Comment by fricklerhandwerk** [comment]
Another pro of `inherit`: tiny performance improvement over attribute selection.

The "which lib does it come from" con is somewhat alleviated by the explicit `inherit`, it just requires the annoying jumping.

In some cases, as in related Python PRs, it's possible to find smaller scopes for the large number of imports. It's a bit of code organisation issue, which can't easily be solved without tearing everything apart, or maybe not at all.

On top of removing `with`, I think this is net positive, however slightly. And arguably one can still shift things around if touching the code shows that it's impractical to work with.

**Review by fricklerhandwerk — APPROVED** [review]

**Comment by maralorn** [comment]
Fair. @sternenseemann, @ncfavier, any objections?

Uff, so many merge conflicts. Sorry.
You will sadly need to fix that up. Can you please create the PR against the haskell-updates branch? Otherwise it will be a nightmare for us. (And good luck fixing and rebasing that without creating an accidental mass ping.)

**Comment by philiptaron** [comment]
I’ll take that as a challenge. I’ll gladly rebase on Haskell updates
sometime today or tomorrow.

Philip (on mobile)

**Comment by sternenseemann** [comment]
I don't really see the benefit. If you've ever looked at the file, it is pretty clear that special functions are in scope that are documented e.g. in the nixpkgs manual (at last!). Maintaining explicit import lists is annoying, prevents overrides from being copied between the files without any problems and will probably lead to merge conflicts. It also doesn't make it any clearer where the functions are coming from since `haskellLib` is a fake name that doesn't correspond to any attributes.

> Another pro of `inherit`: tiny performance improvement over attribute selection.

It makes no difference.

```console
$ hyperfine -L r 3f7b7fa16cfce9768c0f6cbdf88fb4291c3ac633,5e5b589095c19b46923a72d5f05a0d419bcd4616 --setup 'git checkout {r}' "nix-env -qaP -f . -A haskellPackages --out-path"
Benchmark 1: nix-env -qaP -f . -A haskellPackages --out-path (r = 3f7b7fa16cfce9768c0f6cbdf88fb4291c3ac633)
  Time (mean ± σ):     17.906 s ±  0.078 s    [User: 16.968 s, System: 0.930 s]
  Range (min … max):   17.790 s … 18.027 s    10 runs
 
Benchmark 2: nix-env -qaP -f . -A haskellPackages --out-path (r = 5e5b589095c19b46923a72d5f05a0d419bcd4616)
  Time (mean ± σ):     17.977 s ±  0.090 s    [User: 17.049 s, System: 0.920 s]
  Range (min … max):   17.837 s … 18.120 s    10 runs
 
Summary
  nix-env -qaP -f . -A haskellPackages --out-path (r = 3f7b7fa16cfce9768c0f6cbdf88fb4291c3ac633) ran
    1.00 ± 0.01 times faster than nix-env -qaP -f . -A haskellPackages --out-path (r = 5e5b589095c19b46923a72d5f05a0d419bcd4616)
```

**Comment by philiptaron** [comment]
Is it expected that `haskell-updates` fails to `nixpkgs-review`? I see this without any of my refactors.

```
$ nixpkgs-review rev upstream/haskell-updates -b haskell-updates
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs haskell-updates:refs/nixpkgs-review/0
$ git worktree add /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs 1125177568a762024d594d04e4924521511134b3
Preparing worktree (detached HEAD 1125177568a7)
Updating files: 100% (40649/40649), done.
HEAD is now at 1125177568a7 haskellPackages: regenerate package set based on current config
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f <nixpkgs> --nix-path nixpkgs=/home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs nixpkgs-overlays=/tmp/tmpmva809i0 -qaP --xml --out-path --show-trace --no-allow-import-from-derivation
error:
       … while querying the derivation named 'changelog-d-0.1-git-2816ddb'

       … while calling the 'getAttr' builtin

         at /builtin/derivation.nix:19:19: (source not available)

       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'changelog-d-0.1-git-2816ddb'
         whose name attribute is located at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'buildInputs' of derivation 'changelog-d-0.1-git-2816ddb'

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:378:7:

          377|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          378|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          379|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       … while evaluating derivation 'cabal-install-parsers-0.6.1.1'
         whose name attribute is located at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'propagatedBuildInputs' of derivation 'cabal-install-parsers-0.6.1.1'

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:385:7:

          384|       depsHostHostPropagated      = elemAt (elemAt propagatedDependencies 1) 0;
          385|       propagatedBuildInputs       = elemAt (elemAt propagatedDependencies 1) 1;
             |       ^
          386|       depsTargetTargetPropagated  = elemAt (elemAt propagatedDependencies 2) 0;

       … while calling anonymous lambda

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:321:13:

          320|       (map (drv: getDev drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHostPropagated" depsHostHostPropagated))
          321|       (map (drv: getDev drv.__spliced.hostTarget or drv) (checkDependencyList "propagatedBuildInputs" propagatedBuildInputs))
             |             ^
          322|     ]

       … from call site

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:321:18:

          320|       (map (drv: getDev drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHostPropagated" depsHostHostPropagated))
          321|       (map (drv: getDev drv.__spliced.hostTarget or drv) (checkDependencyList "propagatedBuildInputs" propagatedBuildInputs))
             |                  ^
          322|     ]

       … while calling 'getOutput'

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/lib/attrsets.nix:1762:23:

         1761|   */
         1762|   getOutput = output: pkg:
             |                       ^
         1763|     if ! pkg ? outputSpecified || ! pkg.outputSpecified

       … while calling anonymous lambda

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/lib/lists.nix:334:29:

          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                             ^
          335|

       … from call site

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/lib/lists.nix:334:32:

          333|   */
          334|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                ^
          335|

       … while calling anonymous lambda

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:276:15:

          275|     imap1
          276|       (index: dep:
             |               ^
          277|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep

       … from call site

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:277:12:

          276|       (index: dep:
          277|         if isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |            ^
          278|         else if isList dep then checkDependencyList' ([index] ++ positions) name dep

       … while calling 'isDerivation'

         at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/lib/attrsets.nix:1249:5:

         1248|   isDerivation =
         1249|     value: value.type or null == "derivation";
             |     ^
         1250|

       error: attribute 'Cabal-syntax_3_10_2_0' missing

       at /home/philip/.cache/nixpkgs-review/rev-1125177568a762024d594d04e4924521511134b3/nixpkgs/pkgs/development/haskell-modules/configuration-common.nix:1515:20:

         1514|   cabal-install-parsers = doJailbreak (dontCheck (super.cabal-install-parsers.override {
         1515|     Cabal-syntax = self.Cabal-syntax_3_10_2_0;
             |                    ^
         1516|   }));
       Did you mean Cabal-syntax_3_10_3_0?
```

**Comment by philiptaron** [comment]
Seeing there was controversy about the shortening and making consistent names from `lib`, `pkgs.lib`, and `builtins`, I have omitted those changes from this PR. 

I'm more interested in seeing the removal of top-level with than I am pained by the inconsistency of `lib`, `pkgs.lib`, `builtins`, and so forth.

**Comment by maralorn** [comment]
You just encountered the branch at a bad time. We will need a moment to fix eval.

@sternenseemann but are you okay with getting rid of the with? At least in my setup with destroys unbound variable warnings and therefore getting rid of it would be quite beneficial.

Am 31. März 2024 23:06:57 MESZ schrieb Philip Taron ***@***.***>:

**Comment by philiptaron** [comment]
I just rebased on `haskell-updates` a few minutes ago and successfully completed `nixpkgs-review rev HEAD -b haskell-updates`.

**Comment by philiptaron** [comment]
Rebased on `haskell-updates`. ☝🏻

**Comment by sternenseemann** [comment]
> Is it expected that `haskell-updates` fails to `nixpkgs-review`?

Some revisions may fail to evaluate or have massive regressions, yes. Eval errors usually get cleaned up quite quickly.

**Comment by sternenseemann** [comment]
@maralorn well, all I wrote was about the `with`…

**Comment by maralorn** [comment]
@philiptaron since a majority of Haskell maintainers seems to be against this change (@ncfavier upvoted the veto by @sternenseemann). I fear there is not much point in pursuing this further. I am really sorry for the work you invested. I could maybe have anticipated this before I made you redo this PR basically twice, but I didn’t. For that I am also sorry.

**Comment by philiptaron** [comment]
I'm pretty disappointed by that result. There's a pretty broad consensus that top-level `with`, by virtue of turning off unbound variable warnings, makes it more likely to cause errors which are detected quite a bit later than they would be otherwise.

I agree with @sternenseemann on these points.

1. Maintaining explicit import lists is annoying
2. Different variables in scope prevents overrides from being copied between the files without any problems.
3. Having lists leads to merge conflicts.

In my estimation, the benefit accrued from getting unbound variable warnings is worth those downsides (which aren't particularly horrendous!).

If top-level `with` meant that I got good errors about unbound variables, I would recommend it.

I disagree or would rephrase these points:

* It is pretty clear that special functions are in scope.
* Eliminating the `with` also doesn't make it any clearer where the functions are coming from since `haskellLib` is a fake name that doesn't correspond to any attributes.

On the first point, sure: `with a` means that the things in the attrset `a` are scope. What's unclear is that only some of those things are coming from attrset `a`, and others are coming from arguments, let bindings, etc. It also allows through typos and other common mistakes. The so-called clarity of bringing those names into scope, as I see it, doesn't pay for the cost. 

And whether it _is_ clear which special functions are in scope is called into question by the second point. If `haskellLib` is a fake name that doesn't correspond to attributes, then how can one know the functions that it brings into scope? That seems a self-defeating argument to me.

For what it's worth, `vim -t haskellLib` then going through the tags generated by [`nix-doc`](https://github.com/lf-/nix-doc) brought me enough insight to find the symbols from that attrset.

You all are the gatekeepers here, though. I'm going to leave this open, in the hopes that minds are changed over time.
